File: PsyScopeEditor/PSPluginLoader.m
//
//  PSPluginLoader.m
//  PsyScope Editor
//
//  Created by James on 09/07/2014.
//


#import "PSPluginLoader.h"

@implementation PSPluginLoader


- (id)init {
    self = [super init];
    pluginClasses = [[NSMutableDictionary alloc] init];
   
    bundlePaths = [[NSMutableDictionary alloc] init];

    
    tools = [[NSMutableArray alloc] init];
    attributes = [[NSMutableArray alloc] init];
    events = [[NSMutableArray alloc] init];

    windowViews = [[NSMutableArray alloc] init];
    actions = [[NSMutableArray alloc] init];
    conditions = [[NSMutableArray alloc] init];
    
    //get plugin names and remember which PluginInterface they were associated with
    
    NSString* folderPath = [[NSBundle mainBundle] builtInPlugInsPath];
    if (folderPath) {
        //get all plugins in folder path
        NSEnumerator* enumerator = [[NSBundle pathsForResourcesOfType:@"psyplugin" inDirectory:folderPath] objectEnumerator];
        NSString* pluginPath;
        while ((pluginPath = [enumerator nextObject])) {
            //for each plugin get the principle class and use the class methods to get the actual plugins
            NSBundle* pluginBundle = [NSBundle bundleWithPath:pluginPath];
            if (pluginBundle) {
                NSDictionary* pluginDict = [pluginBundle infoDictionary];
                NSString* pluginName = [pluginDict objectForKey:@"NSPrincipalClass"];
                if (pluginName) {
                    Class pluginClass = NSClassFromString(pluginName);
                    if (!pluginClass) {
                        pluginClass = [pluginBundle principalClass];
                        if ([pluginClass conformsToProtocol:@protocol(PSPluginInterface)] && [pluginClass isKindOfClass:[NSObject class]]) {
                            //now get all class names from the interface, and instantiate objects with the class names as keys
                            
                            //TODO refactor
                            
                            NSArray* names = [pluginClass pluginsFor:PSPluginTypeTool];
                            [tools addObjectsFromArray:names];
                            for (id object in names) {
                                //this should just be strings in here
                                if ([object isKindOfClass:[NSString class]]) {
                                    NSString* name = (NSString *)object;
                                    [pluginClasses setObject:[pluginClass getPSExtensionClass:name] forKey:name];
                              
                                    [bundlePaths setObject:pluginPath forKey:name];
                                }
                                
                            }
                            
                            //now for attributes
                            
                            names = [pluginClass pluginsFor:PSPluginTypeAttribute];
                            [attributes addObjectsFromArray:names];
                            for (id object in names) {
                                //this should just be strings in here
                                if ([object isKindOfClass:[NSString class]]) {
                                    NSString* name = (NSString *)object;
                                    [pluginClasses setObject:[pluginClass getPSExtensionClass:name] forKey:name];
                                 
                                    [bundlePaths setObject:pluginPath forKey:name];
                                }
                                
                            }
                            
                            //now for events
                            
                            names = [pluginClass pluginsFor:PSPluginTypeEvent];
                            [events addObjectsFromArray:names];
                            for (id object in names) {
                                //this should just be strings in here
                                if ([object isKindOfClass:[NSString class]]) {
                                    NSString* name = (NSString *)object;
                                    [pluginClasses setObject:[pluginClass getPSExtensionClass:name] forKey:name];
                                 
                                    [bundlePaths setObject:pluginPath forKey:name];
                                }
                                
                            }
                            
                            //now for window views
                            names = [pluginClass pluginsFor:PSPluginTypeWindowView];
                            [windowViews addObjectsFromArray:names];
                            for (id object in names) {
                                //this should just be strings in here
                                if ([object isKindOfClass:[NSString class]]) {
                                    NSString* name = (NSString *)object;
                                    [pluginClasses setObject:[pluginClass getPSExtensionClass:name] forKey:name];
                            
                                    [bundlePaths setObject:pluginPath forKey:name];
                                }
                                
                            }
                            
                            //now for actions
                            names = [pluginClass pluginsFor:PSPluginTypeAction];
                            [actions addObjectsFromArray:names];
                            for (id object in names) {
                                //this should just be strings in here
                                if ([object isKindOfClass:[NSString class]]) {
                                    NSString* name = (NSString *)object;
                                    [pluginClasses setObject:[pluginClass getPSExtensionClass:name] forKey:name];
                               
                                    [bundlePaths setObject:pluginPath forKey:name];
                                }
                                
                            }
                            
                            //now for conditions
                            names = [pluginClass pluginsFor:PSPluginTypeCondition];
                            [conditions addObjectsFromArray:names];
                            for (id object in names) {
                                //this should just be strings in here
                                if ([object isKindOfClass:[NSString class]]) {
                                    NSString* name = (NSString *)object;
                                    [pluginClasses setObject:[pluginClass getPSExtensionClass:name] forKey:name];
                               
                                    [bundlePaths setObject:pluginPath forKey:name];
                                }
                                
                            }
                            
                            
                        }
                    }
                }
            }
        }
    }
    
    
    return self;
}

//	Scan the application's plug-in folder for plug-ins and try to activate them.

-(NSArray*)getPluginClassNames {

    return [pluginClasses allKeys];
}



-(NSArray*) instantiatePluginsOfType:(PSPluginType)pluginType {


    NSArray* selectedArray;
    
    switch (pluginType) {
        case PSPluginTypeTool:

            selectedArray = tools;
            break;
        case PSPluginTypeAttribute:
     
            selectedArray = attributes;
            break;
        case PSPluginTypeEvent:
        
            selectedArray = events;
            break;
        case PSPluginTypeWindowView:
         
            selectedArray = windowViews;
            break;
        case PSPluginTypeAction:
       
            selectedArray = actions;
            break;
        case PSPluginTypeCondition:
      
            selectedArray = conditions;
            break;
    }
    
    NSMutableArray* return_array = [[NSMutableArray alloc] init];
    
    for(id key in selectedArray) {
        Class extensionClass = [pluginClasses objectForKey: key];
        [return_array addObject:[[extensionClass alloc] init]];
    }
    
    return return_array;
}





                             
-(NSString*) getBundlePathForPSExtension:(NSString*)name {
    NSString* path = (NSString*) [bundlePaths objectForKey:name];
    return path;
}




@end

File: PsyScopeEditor/PSPluginLoader.swift
//
//  PSPluginLoader.swift
//  PsyScopeEditor
//
//  Created by James on 16/10/2015.
//  Copyright © 2015 James. All rights reserved.
//

import Foundation

class PSPluginLoader : NSObject {
    
    
    var pluginClasses :[String : AnyObject] = [:]
    var bundlePaths : [String : String] = [:]
    
    var tools : [PSToolInterface] = []
    var events : [PSEventInterface] = []
    var attributes : [PSAttributeInterface] = []
    var windowViews : [PSWindowViewInterface] = []
    var actions : [PSActionInterface] = []
    var conditions : [PSConditionInterface] = []
    
    var pluginsLoaded : [String] = []
    override init() {
        super.init()
        guard let builtInPlugInsPath = Bundle.main.builtInPlugInsPath else {
            fatalError("Could not get built in plugins path")
        }
        
        //search in inbuilt plugins
        for resourcePath in Bundle.paths(forResourcesOfType: "psyplugin", inDirectory: builtInPlugInsPath) {
            loadPluginsInPath(resourcePath)
            pluginsLoaded.append(resourcePath)
        }
    
        //also search custom path
        let customPath = PSPreferences.pluginPath.stringValue
        for resourcePath in Bundle.paths(forResourcesOfType: "psyplugin", inDirectory: customPath) {
            loadPluginsInPath(resourcePath)
            pluginsLoaded.append(resourcePath)
        }
        
    }
        
    func loadPluginsInPath(_ resourcePath : String) {
        guard let pluginBundle = Bundle(path: resourcePath),
            let pluginClass = pluginBundle.principalClass,
            let pluginInterface = pluginClass as? PSPluginInterface.Type else {
                print("Incorrect setup for plugin at \(resourcePath) - bundle principal class must be PSPluginInterface")
                return
        }
        
        let toolPlugins = setupPluginsFor(PSPluginType.tool, pluginInterface: pluginInterface, resourcePath: resourcePath)
        tools += toolPlugins.map({ $0 as! PSToolInterface })
        
        let eventPlugins = setupPluginsFor(PSPluginType.event, pluginInterface: pluginInterface, resourcePath: resourcePath)
        events += eventPlugins.map({ $0 as! PSEventInterface })
        
        let attributePlugins = setupPluginsFor(PSPluginType.attribute, pluginInterface: pluginInterface, resourcePath: resourcePath)
        attributes += attributePlugins.map({ $0 as! PSAttributeInterface })
        
        let wvPlugins = setupPluginsFor(PSPluginType.windowView, pluginInterface: pluginInterface, resourcePath: resourcePath)
        windowViews += wvPlugins.map({ $0 as! PSWindowViewInterface })
        
        let actionPlugins = setupPluginsFor(PSPluginType.action, pluginInterface: pluginInterface, resourcePath: resourcePath)
        actions += actionPlugins.map({ $0 as! PSActionInterface })
        
        let conditionPlugins = setupPluginsFor(PSPluginType.condition, pluginInterface: pluginInterface, resourcePath: resourcePath)
        conditions += conditionPlugins.map({ $0 as! PSConditionInterface })
        
    }
        
    
    func setupPluginsFor(_ type : PSPluginType, pluginInterface : PSPluginInterface.Type, resourcePath : String) -> [NSObject] {
        let classes = pluginInterface.pluginsFor(type)
        var pluginInstances : [NSObject] = []
        for pluginClass in classes {
            
            let name = String(describing: Swift.type(of: pluginClass))
            let pluginObject = pluginClass.init()
            pluginInstances.append(pluginObject)
            pluginClasses[name] = pluginClass
            bundlePaths[name] = resourcePath
        }
        return pluginInstances
    }
}
